README.md
# hyp-chain
HYP Produse Legate


changelog.txt
*** WooCommerce Chained Products Changelog ***

2019.06.11 - version 2.9.2
 * Fix: Chained Product search not returning all search results (products) in some cases
 * Fix: Welcome page CSS
 * Update: Tested upto WordPress 5.2.1 and WooCommerce 3.6.4
 * Update: POT file

2019.05.07 - version 2.9.1
 * New: WooCommerce 3.6.2 compatible
 * Fix: WooCommerce custom product types did not get chained as chained products in some cases
 * Fix: Chained products fields not showing in admin when adding a new product in some cases
 * Fix: Welcome page CSS added on all admin pages
 * Update: Plugin action links
 * Update: Welcome page
 * Update: Admin notices
 * Update: POT file

2019.03.07 - version 2.9.0
 * New: Allow chained products to be set for all WooCommerce product types (https://docs.woocommerce.com/document/chained-products/#section-20)
 * Fix: Shortcode vanishing from the editor on publish/update of product
 * Fix: Sometimes quantity not updated correctly in existing orders
 * Update: Always show insert shortcode notice on product add/edit page
 * Update: Tested with WordPress 5.1 and latest WooCommerce
 * Update: Plugin action links
 * Update: Welcome page
 * Update: POT file

2019.01.22 - version 2.8.4
 * Fix: Coupon not discounting on Priced Individually Chained Products
 * Update: Tested with PHP 7.3 and latest WooCommerce
 * Update: Welcome page
 * Update: POT file

2018.11.14 - version 2.8.3
 * Fix: Bundle price set to zero on cart page (Compatibility issue with WooCommerce Product Bundles)
 * Fix: Price always displayed with tax on cart page for product having chained items
 * Fix: Don't allow adding a product addon for chained item (Compatibility issue with WooCommerce Product Add-ons) IS#25

2018.06.11 - version 2.8.2
 * New: WooCommerce 3.5.0+ compatible
 * Fix: Fatal error "Call to a member function get_id() on boolean"

2018.10.22 - version 2.8.1
 * Fix: PHP Notice: Undefined variable: override_total
 * Update: POT file

2018.10.15 - version 2.8.0
 * New: Feature to allow chained item to be priced individually (For WooCommerce 3.0.0+)
 * Update: Welcome page
 * Update: POT file

2018.09.12 - version 2.7.0
 * New: WordPress Coding Standards
 * New: Settings page for chained products
 * Fix: Compatibility issue with WooCommerce version < 3.0
 * Update: Doc and help links
 * Update: POT file

2018.07.02 - version 2.6.2
 * New: WooCommerce 3.4 compatible
 * Fix: Chained item showing on-sale tag on shop/product page when the same chained item is present in cart
 * Fix: Deprecation notice on purchase of a subscription product

2018.05.21 - version 2.6.1
 * New: WooCommerce 3.3 compatible
 * Fix: Shortcode vanishing from the editor on publish/update of variable product
 * Fix: Do not accept coupon for on-sale products with chained items when 'Exclude sale items' option is checked
 * Fix: Notice not being displayed when chained items are out of stock and 'Manage Stock?' option is checked
 * Fix: Error notice when product with chained item is removed from cart (for WooCommerce 3.2+)
 * Update: POT file

2018.02.01 - version 2.6.0
 * New: WooCommerce 3.2 compatible
 * Fix: Too many redirect issue on multi-site setup
 * Fix: Fatal error when pulling products having chained item using REST API
 * Fix: Show notice on product page when chained item is out of stock
 * Fix: Select2 search issue with variable product
 * Fix: Incorrect order total when creating order from back-end
 * Fix: Old chained item not getting removed while switching subscription

2017.10.09 - version 2.5.7
 * Fix: Allow adding chained products to cart if back ordering is allowed
 * Fix: Price not set to zero for chained variable product while manually creating order from backend
 * Update: POT file

2017.09.06 - version 2.5.6
 * Fix: Select2 restricting selection of parent product in select2 of other plugins.
 * Update: Suppress 'Error Loading' message while searching for a product

2017.07.26 - version 2.5.5
 * Fix: Don't allow chaining a variation to itself
 * Fix: Chaining issue on order received page

2017.07.24 - version 2.5.4
 * Fix: Compatibility issue with WooCommerce version lower than 3.0

2017.07.10 - version 2.5.3
 * Fix: Fatal error on subscription renewal with WooCommerce Subscriptions - v2.2.8

2017.06.22- version 2.5.2
 * Fix: WooCommerce Bookings emails not being send when Chained Products is active

2017.05.11- version 2.5.1
 * Fix: Search for variable products returning parent variable product
 * Fix: Mini Cart count count sets to zero when 'sa_cp_show_chained_items' filter was used
 * Fix: Compatibility notices and warnings when 'sa_show_chained_item_price' is set to yes

2017.04.07 - version 2.5
 * New: WooCommerce 3.0 compatible
 * New: WooCommerce Subscriptions 2.2.0 compatible
 * Fix: UI issue on order screen for WooCommerce less than 3.0
 * Fix: When hide out of stock items in checked then the parent product should hide too if chained product is out of stock.
 * Fix: Take backorders into consideration for chained products
 * Update: Class structure for frontend code
 * Update: Will now support only WooCommmerce 2.5+
 * Update: POT file

2016.11.10 - version 2.4.4
 * Fix: Mini Cart count not getting updated when 'sa_cp_show_chained_items' filter was used
 * Fix: Incorrect stock count for variable product
 * Fix: Variable product title not getting parsed correctly when used with different language
 * Update: Textdomain correction
 * Update: POT file

2016.09.19 - version 2.4.3
 * Fix: Chained child items getting added twice, after clicking Pay button from Order

2016.06.20 - version 2.4.2
 * New: Tested upto WooCommerce 2.6+
 * Fix: Housekeeping feature not working for variable products with chained items
 * Update: POT file

2016.05.12 - version 2.4.1
 * Fix: Updating chained grouped products quantities in cart, which lead to removal of chained products from cart
 * Fix: Removal of some warnings related to database queries

2016.04.27 - version 2.4
 * New: Housekeeping feature for chained products

woocommerce-chained-products.php
<?php
/**
 * Plugin Name: HYP Chain
 * Plugin URI: https://github.com/hypericumimpex/hyp-chain/
 * Description: Produse Legate.
 * Version: 2.9.2
 * Author: StoreApps
 * Author URI: https://github.com/hypericumimpex/hyp-chain/
 * Developer: StoreApps
 * Developer URI: https://github.com/hypericumimpex/
 * Requires at least: 4.1
 * Tested up to: 5.2.1
 * WC requires at least: 2.5.0
 * WC tested up to: 3.6.4
 * Text Domain: woocommerce-chained-products
 * Domain Path: /languages/
 * Woo: 18687:cc6e246e495745db10f9f7fddc5aa907
 * Copyright (c) 2012-2019 WooCommerce, StoreApps All rights reserved.
 * License: GNU General Public License v3.0
 * License URI: http://www.gnu.org/licenses/gpl-3.0.html
 *
 * @package woocommerce-chained-products
 */

/**
 * Required functions
 */
if ( ! function_exists( 'woothemes_queue_update' ) ) {
	require_once 'woo-includes/woo-functions.php';
}

/**
 * Plugin updates
 */
woothemes_queue_update( plugin_basename( __FILE__ ), 'cc6e246e495745db10f9f7fddc5aa907', '18687' );

/**
 * Function to set transient on plugin activation
 */
function chained_product_activate() {
	if ( ! is_network_admin() && ! isset( $_GET['activate-multi'] ) ) { // WPCS: input var ok, CSRF ok.
		set_transient( '_chained_products_activation_redirect', 1, 30 );
	}
}

register_activation_hook( __FILE__, 'chained_product_activate' );

if ( is_woocommerce_active() ) {

	/**
	 * Function to initiate Chained Products & its functionality
	 */
	function initialize_chained_products() {
		global $wc_cp;

		if ( ! defined( 'WC_CP_PLUGIN_DIRNAME' ) ) {
			define( 'WC_CP_PLUGIN_DIRNAME', dirname( plugin_basename( __FILE__ ) ) );
		}

		if ( ! defined( 'WC_CP_PLUGIN_FILE' ) ) {
			define( 'WC_CP_PLUGIN_FILE', __FILE__ );
		}

		include_once 'includes/class-wc-chained-products.php';

		$wc_cp = new WC_Chained_Products();
	}

	add_action( 'plugins_loaded', 'initialize_chained_products' );
}


includes/class-wc-cp-admin-notices.php
<?php
/**
 * Class to handle display of Chained Products admin notices
 *
 * @package woocommerce-chained-products/includes/
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}

if ( ! class_exists( 'WC_CP_Admin_Notices' ) ) {

	/**
	 * Class to handle display of Chained Products review notice.
	 *
	 * @author StoreApps
	 */
	class WC_CP_Admin_Notices {

		/**
		 * The msg heading for review.
		 *
		 * @var string
		 */
		public $msg = '';

		/**
		 * Constructor
		 */
		public function __construct() {
			add_filter( 'plugin_action_links_' . plugin_basename( WC_CP_PLUGIN_FILE ), array( $this, 'plugin_action_links' ) );

			add_action( 'admin_notices', array( $this, 'admin_notice_sa_needs_wc_25_above' ) );
			add_action( 'admin_init', array( $this, 'cp_dismiss_admin_notice' ), 12 );

			add_action( 'admin_notices', array( $this, 'sa_cp_show_review_notice' ) );
			add_action( 'admin_init', array( $this, 'sa_cp_update_notice_action' ) );
		}

		/**
		 * Function to handle WC compatibility related function call from appropriate class.
		 *
		 * @param string $function_name Function to call.
		 * @param array  $arguments Array of arguments passed while calling $function_name.
		 * @return mixed Result of function call.
		 */
		public function __call( $function_name, $arguments = array() ) {

			if ( ! is_callable( 'Chained_Products_WC_Compatibility', $function_name ) ) {
				return;
			}

			if ( ! empty( $arguments ) ) {
				return call_user_func_array( 'Chained_Products_WC_Compatibility::' . $function_name, $arguments );
			} else {
				return call_user_func( 'Chained_Products_WC_Compatibility::' . $function_name );
			}
		}

		/**
		 * Function to add more action on plugins page
		 *
		 * @param array $links Existing links.
		 * @return array $links
		 */
		public function plugin_action_links( $links ) {
			$args = array(
				'page'    => 'wc-settings',
				'tab'     => 'products',
				'section' => 'wc_chained_products',
			);

			$settings_url  = add_query_arg( $args, admin_url( 'admin.php' ) );
			$about_url     = add_query_arg( 'page', 'cp-about', admin_url( 'admin.php' ) );
			$shortcode_url = add_query_arg( 'page', 'cp-shortcode', admin_url( 'admin.php' ) );

			$action_links = array(
				'shortcode' => '<a target="_blank" href="' . esc_url( $shortcode_url ) . '">' . __( 'Shortcode', 'woocommerce-chained-products' ) . '</a>',
				'settings'  => '<a target="_blank" href="' . esc_url( $settings_url ) . '">' . __( 'Settings', 'woocommerce-chained-products' ) . '</a>',
				'docs'      => '<a target="_blank" href="' . esc_url( 'https://docs.woocommerce.com/document/chained-products/' ) . '">' . __( 'Docs', 'woocommerce-chained-products' ) . '</a>',
				'support'   => '<a target="_blank" href="' . esc_url( 'https://woocommerce.com/my-account/create-a-ticket/' ) . '">' . __( 'Support', 'woocommerce-chained-products' ) . '</a>',
				'review'    => '<a target="_blank" href="' . esc_url( 'https://woocommerce.com/products/chained-products/#comments' ) . '">' . __( 'Review', 'woocommerce-chained-products' ) . '</a>',
			);

			return array_merge( $action_links, $links );
		}

		/**
		 * Function to show admin notice that Chained Products works with WC 2.5.0+
		 */
		public function admin_notice_sa_needs_wc_25_above() {

			if ( ! $this->is_wc_gte_25() ) {
				?>
				<div class="updated error">
					<p>
					<?php
						printf(
							'<strong>' . esc_html__( 'Important - ', 'woocommerce-chained-products' ) . '</strong>' . esc_html__( 'WooCommerce Chained Products plugin is active but it will only work with WooCommerce 2.5+. ', 'woocommerce-chained-products' ) . '<a href="%s">' . esc_html__( 'Please update WooCommerce to the latest version', 'woocommerce-chained-products' ) . '</a>',
							esc_url( admin_url( 'plugins.php?' ) )
						);
					?>
					</p>
				</div>
				<?php
			}

		}

		/**
		 * Function to dismiss admin notice
		 */
		public function cp_dismiss_admin_notice() {
			if ( isset( $_GET['cp_dismiss_admin_notice'] ) && '1' === $_GET['cp_dismiss_admin_notice'] && isset( $_GET['option_name'] ) ) { // WPCS: CSRF ok, input var ok.
				$option_name = sanitize_text_field( wp_unslash( $_GET['option_name'] ) ); // WPCS: CSRF ok, input var ok.
				update_option( $option_name, 'no' );

				$referer = wp_get_referer();
				wp_safe_redirect( $referer );
				exit();
			}
		}

		/**
		 * Shows review notice
		 */
		public function sa_cp_show_review_notice() {
			global $post;
			$is_review_page = ( isset( $post->post_type ) && ( 'shop_order' === $post->post_type || 'product' === $post->post_type ) ) ? true : false; // WPCS: CSRF ok.

			if ( true === $this->may_be_show_review_notice() && true === $is_review_page ) {
				$this->cp_show_review_notice();
			}
		}

		/**
		 * Function to decide whether to show the display notice or not.
		 *
		 * @return bool $show_notice
		 */
		public function may_be_show_review_notice() {
			$show_notice = false;

			$cp_show_review_notice = get_option( 'cp_show_review_notice', 'yes' );

			if ( 'no' !== $cp_show_review_notice ) {
				$this->msg = __( 'Glad you are using', 'woocommerce-chained-products' ) . '&nbsp;<strong>' . __( 'WooCommerce Chained Products.', 'woocommerce-chained-products' ) . '</strong><br>'; // Default msg.

				if ( 'yes' === $cp_show_review_notice ) {
					if ( true === $this->orders_has_chained_item() ) {
						$show_notice = true;
						$this->msg   = '<strong>' . __( 'Congratulations!', 'woocommerce-chained-products' ) . '</strong>&nbsp;' . esc_html__( 'You have successfully sold a product using', 'woocommerce_chained_product' ) . '&nbsp;<strong>' . __( 'WooCommerce Chained Products', 'woocommerce_chained_product' ) . '</strong>&nbsp;' . esc_html__( 'plugin.', 'woocommerce-chained-products' ) . '<br>';
					} elseif ( true === $this->products_has_chained_item() ) {
						$show_notice = true;
					}
				} elseif ( time() >= absint( $cp_show_review_notice ) ) {
					$show_notice = true;
				}
			}

			return $show_notice;
		}

		/**
		 * Function to check if orders contains chained items or not.
		 *
		 * @return bool $has_chained_items
		 */
		public function orders_has_chained_item() {
			global $wpdb;

			$order_count = wp_cache_get( 'wc_cp_order_count', 'woocommerce_chained_product' );

			if ( false === $order_count ) {
				$order_count = $wpdb->get_var(
					$wpdb->prepare(
						"SELECT count( DISTINCT order_id )
					FROM {$wpdb->prefix}woocommerce_order_items AS o
					JOIN {$wpdb->prefix}woocommerce_order_itemmeta AS oi ON ( o.order_item_id = oi.order_item_id AND oi.meta_key = %s )
					JOIN {$wpdb->prefix}posts AS p ON (o.order_id = p.ID AND p.post_type = %s AND p.post_status = %s )",
						'_chained_product_of',
						'shop_order',
						'wc-completed'
					)
				); // WPCS: db call ok.

				wp_cache_set( 'wc_cp_order_count', $order_count, 'woocommerce_chained_product' );
			}

			$has_chained_items = ( 0 < $order_count ) ? true : false;

			return $has_chained_items;
		}

		/**
		 * Function to check if products contains chained items or not.
		 *
		 * @return bool $has_chained_items
		 */
		public function products_has_chained_item() {
			global $wpdb;

			$product_count = wp_cache_get( 'wc_cp_product_count', 'woocommerce_chained_product' );

			if ( false === $product_count ) {
				$product_count = $wpdb->get_var(
					$wpdb->prepare(
						"SELECT count( ID )
					FROM {$wpdb->prefix}posts AS p
					JOIN {$wpdb->prefix}postmeta AS pm
					ON ( p.ID = pm.post_id AND pm.meta_key = %s AND p.post_status = %s AND ( p.post_type = %s || p.post_type = %s ) AND DATEDIFF( now(), p.post_date ) > 30 )",
						'_chained_product_detail',
						'publish',
						'product',
						'product_varitation'
					)
				); // WPCS: db call ok.

				wp_cache_set( 'wc_cp_product_count', $product_count, 'woocommerce_chained_product' );
			}

			$has_chained_items = ( 0 < $product_count ) ? true : false;

			return $has_chained_items;
		}


		/**
		 * 5 star review notice content.
		 */
		public function cp_show_review_notice() {
			?>
			<div id="wc_cp_review_notice" class="notice updated fade" style="background-color: mintcream;">
				<div class="wc_cp_review_notice_action" style="float: right;padding: 0.5em 0;text-align: right;font-size: 0.9em;">
					<a href="?cp_notice_action=remind" class="wc_cp_review_notice_remind"><?php echo esc_html__( 'Remind me after a month', 'woocommerce-chained-products' ); ?></a><br>
					<a href="?cp_notice_action=dismiss" class="wc_cp_review_notice_remove"><?php echo esc_html__( 'Never show again', 'woocommerce-chained-products' ); ?></a>
				</div>
				<p>
					<?php echo $this->msg . esc_html__( 'It would be great if you ', 'woocommerce-chained-products' ) . ' <a target="__blank" href="' . esc_url( 'https://woocommerce.com/products/chained-products/#comments' ) . '">' . esc_html__( 'give us a 5-star rating', 'woocommerce-chained-products' ) . '</a>&nbsp;' . esc_html__( 'of your experience. Thanking you in advance 😊', 'woocommerce-chained-products' ); // WPCS: XSS ok. ?>
				</p>
			</div>
			<?php
		}

		/**
		 * Function to update notice action on click of Dismiss/Remind me Later.
		 */
		public function sa_cp_update_notice_action() {
			$action = ( isset( $_GET['cp_notice_action'] ) ) ? wc_clean( wp_unslash( $_GET['cp_notice_action'] ) ) : ''; // WPCS: CSRF ok, sanitization ok, input var ok.

			if ( ! empty( $action ) ) {
				switch ( $action ) {
					case 'remind':
						$option_value = strtotime( '+1 month' );
						break;
					case 'dismiss':
						$option_value = 'no';
						break;
					default:
						$option_value = 'no';
				}

				update_option( 'cp_show_review_notice', $option_value );

				$referer = wp_get_referer();
				wp_safe_redirect( $referer );
				exit();
			}
		}
	}
}

new WC_CP_Admin_Notices();
